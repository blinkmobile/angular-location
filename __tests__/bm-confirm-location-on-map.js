/* @flow */
'use strict'

// see: https://docs.angularjs.org/guide/unit-testing
// see: http://facebook.github.io/jest/docs/tutorial-jquery.html

const ng = require('angular')
require('angular-mocks')

jest.mock('@blinkmobile/geolocation', () => ({
  getCurrentPosition: jest.fn(() => Promise.resolve({
    coords: { latitude: 10, longitude: 10 }
  }))
}))

const fakes = require('./helpers/fakes.js')
const mod = require('../lib/components/bm-confirm-location-on-map.js')
require('../lib/index.js')

beforeEach(ng.mock.module('bmLocation'))

let $compile, $rootScope
beforeEach(ng.mock.inject((_$compile_, _$rootScope_) => {
  $compile = _$compile_
  $rootScope = _$rootScope_
}))

test('no attributes', () => {
  expect(() => {
    $rootScope.googleMapsApiKey = 'secret'
    const html = `
    <bm-confirm-location-on-map></bm-confirm-location-on-map>
    `
    $compile(html)($rootScope)
  }).toThrow()
})

test('$rootScope.googleMapsApiKey [ng-model]', () => {
  $rootScope.googleMapsApiKey = 'secret'
  $rootScope.coords = null
  const html = `
  <bm-confirm-location-on-map
    ng-model="coords"
  ></bm-confirm-location-on-map>
  `

  const element = $compile(html)($rootScope)
  $rootScope.$digest()
  expect(element.html()).toMatchSnapshot()
})

test('$rootScope.googleMapsApiKey [ng-disabled] [ng-model]', () => {
  $rootScope.googleMapsApiKey = 'secret'
  $rootScope.coords = null
  const html = `
  <bm-confirm-location-on-map
    ng-disabled="ng-disabled"
    ng-model="coords"
  ></bm-confirm-location-on-map>
  `

  const element = $compile(html)($rootScope)
  $rootScope.$digest()
  expect(element.html()).toMatchSnapshot()
})

test('$rootScope.googleMapsApiKey [ng-model] [ng-readonly]', () => {
  $rootScope.googleMapsApiKey = 'secret'
  $rootScope.coords = null
  const html = `
  <bm-confirm-location-on-map
    ng-readonly="ng-readonly"
    ng-model="coords"
  ></bm-confirm-location-on-map>
  `

  const element = $compile(html)($rootScope)
  $rootScope.$digest()
  expect(element.html()).toMatchSnapshot()
})

test('$rootScope.googleMapsApiKey [ng-disabled] [ng-model] [ng-readonly]', () => {
  $rootScope.googleMapsApiKey = 'secret'
  $rootScope.coords = null
  const html = `
  <bm-confirm-location-on-map
    ng-disabled="ng-disabled"
    ng-model="coords"
    ng-readonly="ng-readonly"
  ></bm-confirm-location-on-map>
  `

  const element = $compile(html)($rootScope)
  $rootScope.$digest()
  expect(element.html()).toMatchSnapshot()
})

test('controller constructor', () => {
  const ctrl = new mod.BmConfirmLocationOnMapController({ $apply: () => {} })
  expect(ctrl).toBeDefined()
})

test('onEdit() does not trigger geolocation if coords is set', () => {
  const ctrl = new mod.BmConfirmLocationOnMapController({ $apply: () => {} })
  const { getCurrentPosition } = require('@blinkmobile/geolocation')

  ctrl.onChange({ latitude: 10, longitude: 10 })
  ctrl.onEdit()

  expect((getCurrentPosition /* : Function */).mock.calls.length).toBe(0)
})

test('onEdit() does trigger geolocation if coords is set', () => {
  const ctrl = new mod.BmConfirmLocationOnMapController({ $apply: () => {} })
  const { getCurrentPosition } = require('@blinkmobile/geolocation')

  ctrl.onChange(null)
  ctrl.onEdit()

  expect((getCurrentPosition /* : Function */).mock.calls.length).toBe(1)
})

test('$onChanges correctly copies coords', () => {
  const ctrl = new mod.BmConfirmLocationOnMapController({ $apply: () => {} })

  const FAKE_COORDS = { latitude: 20, longitude: 20 }
  ctrl.ngModel = Object.assign(fakes.fakeNgModel(), { $viewValue: FAKE_COORDS })

  ctrl.$onChanges()

  expect(ctrl.coords).toMatchObject(FAKE_COORDS)
})
