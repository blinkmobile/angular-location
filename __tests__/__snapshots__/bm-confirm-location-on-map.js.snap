// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$rootScope.googleMapsApiKey [ng-disabled] [ng-model] [ng-readonly] 1`] = `
"
  <div class=\\"bm-location\\">
    <bm-location-on-map coords=\\"$ctrl.coords\\" ng-disabled=\\"true\\" ng-readonly=\\"true\\" on-change=\\"$ctrl.onChange(value)\\" class=\\"ng-isolate-scope\\" readonly=\\"readonly\\" disabled=\\"disabled\\">
  <!-- ngIf: $ctrl.ngDisabled --><bm-static-location-on-map ng-if=\\"$ctrl.ngDisabled\\" coords=\\"$ctrl.coords\\" class=\\"ng-scope ng-isolate-scope\\">
  <figure class=\\"bm-location__map-container\\">
    <img class=\\"bm-location__map\\" ng-src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\" alt=\\"map of unspecified location\\" height=\\"300\\" width=\\"300\\" src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\">
  </figure>
</bm-static-location-on-map><!-- end ngIf: $ctrl.ngDisabled -->
  <!-- ngIf: !$ctrl.ngDisabled -->
</bm-location-on-map>

    <!-- ngIf: !$ctrl.ngDisabled && !$ctrl.ngReadonly -->
  </div>
"
`;

exports[`$rootScope.googleMapsApiKey [ng-disabled] [ng-model] 1`] = `
"
  <div class=\\"bm-location\\">
    <bm-location-on-map coords=\\"$ctrl.coords\\" ng-disabled=\\"true\\" ng-readonly=\\"false\\" on-change=\\"$ctrl.onChange(value)\\" class=\\"ng-isolate-scope\\" disabled=\\"disabled\\">
  <!-- ngIf: $ctrl.ngDisabled --><bm-static-location-on-map ng-if=\\"$ctrl.ngDisabled\\" coords=\\"$ctrl.coords\\" class=\\"ng-scope ng-isolate-scope\\">
  <figure class=\\"bm-location__map-container\\">
    <img class=\\"bm-location__map\\" ng-src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\" alt=\\"map of unspecified location\\" height=\\"300\\" width=\\"300\\" src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\">
  </figure>
</bm-static-location-on-map><!-- end ngIf: $ctrl.ngDisabled -->
  <!-- ngIf: !$ctrl.ngDisabled -->
</bm-location-on-map>

    <!-- ngIf: !$ctrl.ngDisabled && !$ctrl.ngReadonly -->
  </div>
"
`;

exports[`$rootScope.googleMapsApiKey [ng-model] [ng-readonly] 1`] = `
"
  <div class=\\"bm-location\\">
    <bm-location-on-map coords=\\"$ctrl.coords\\" ng-disabled=\\"true\\" ng-readonly=\\"true\\" on-change=\\"$ctrl.onChange(value)\\" class=\\"ng-isolate-scope\\" readonly=\\"readonly\\" disabled=\\"disabled\\">
  <!-- ngIf: $ctrl.ngDisabled --><bm-static-location-on-map ng-if=\\"$ctrl.ngDisabled\\" coords=\\"$ctrl.coords\\" class=\\"ng-scope ng-isolate-scope\\">
  <figure class=\\"bm-location__map-container\\">
    <img class=\\"bm-location__map\\" ng-src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\" alt=\\"map of unspecified location\\" height=\\"300\\" width=\\"300\\" src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\">
  </figure>
</bm-static-location-on-map><!-- end ngIf: $ctrl.ngDisabled -->
  <!-- ngIf: !$ctrl.ngDisabled -->
</bm-location-on-map>

    <!-- ngIf: !$ctrl.ngDisabled && !$ctrl.ngReadonly -->
  </div>
"
`;

exports[`$rootScope.googleMapsApiKey [ng-model] 1`] = `
"
  <div class=\\"bm-location\\">
    <bm-location-on-map coords=\\"$ctrl.coords\\" ng-disabled=\\"true\\" ng-readonly=\\"false\\" on-change=\\"$ctrl.onChange(value)\\" class=\\"ng-isolate-scope\\" disabled=\\"disabled\\">
  <!-- ngIf: $ctrl.ngDisabled --><bm-static-location-on-map ng-if=\\"$ctrl.ngDisabled\\" coords=\\"$ctrl.coords\\" class=\\"ng-scope ng-isolate-scope\\">
  <figure class=\\"bm-location__map-container\\">
    <img class=\\"bm-location__map\\" ng-src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\" alt=\\"map of unspecified location\\" height=\\"300\\" width=\\"300\\" src=\\"https://maps.googleapis.com/maps/api/staticmap?key=secret&amp;scale=2&amp;size=300x300&amp;zoom=0&amp;center=0%2C0\\">
  </figure>
</bm-static-location-on-map><!-- end ngIf: $ctrl.ngDisabled -->
  <!-- ngIf: !$ctrl.ngDisabled -->
</bm-location-on-map>

    <!-- ngIf: !$ctrl.ngDisabled && !$ctrl.ngReadonly --><div class=\\"bm-location__button-container ng-scope\\" ng-if=\\"!$ctrl.ngDisabled &amp;&amp; !$ctrl.ngReadonly\\">

      <!-- ngIf: $ctrl.isEditing -->
      <!-- ngIf: $ctrl.isEditing -->
      <!-- ngIf: $ctrl.isEditing -->

      <!-- ngIf: !$ctrl.isEditing --><button type=\\"button\\" class=\\"bm-location__button bm-location__button-clear ng-scope\\" ng-if=\\"!$ctrl.isEditing\\" ng-click=\\"$ctrl.onClear()\\">Clear</button><!-- end ngIf: !$ctrl.isEditing -->
      <!-- ngIf: !$ctrl.isEditing --><button type=\\"button\\" class=\\"bm-location__button bm-location__button-edit ng-scope\\" ng-if=\\"!$ctrl.isEditing\\" ng-click=\\"$ctrl.onEdit()\\">Edit
      </button><!-- end ngIf: !$ctrl.isEditing -->

    </div><!-- end ngIf: !$ctrl.ngDisabled && !$ctrl.ngReadonly -->
  </div>
"
`;
